Create full_message & full_message_format validation option [Feature Request]

AlexB52 opened the pull request:
### Summary

Currently, the default format for a full error message is `'%{attribute} %{message}'` this can only be changed with the help of locales and i18n for both ActiveModels & ActiveRecords. 

This new feature adds the ability to format a full error message of an attribute without locales files. It can now be used on any classes that include `ActiveModel::Validations` and fully formatted error messages can be displayed consistently on web views as:

* a group with `errors.full_messages` (on top of a form like the scaffold generator for example)
* next to their appropriate form field with `errors.where(:attribute_name)`. 

Additionally, It seems that the format required from people is either `'%{attribute} %{message}'` or `'%{message}'` this patch  adds a `full_message` option which is a shortcut that uses the `full_message_format: '%{message}'` option introduced.

**What do people think?**

**_I'm happy to update documentation and changelog if this is accepted. I will need guidance about where would be the appropriate place for documentation updates as I'm not sure._** 

### Usage

For apps using locales and i18n nothing has changed. For apps that don't rely much on i18n, devs can format full error messages in the model definition.

Here is a description of the new behaviour, existing behaviour still applies.

```ruby
class Person
  include ActiveModel::Model
  attr_accessor :age, :name

  validates :name, presence: { full_message: 'A person name cannot be blank' }
  validates :age, numericality: {
    greater_than_or_equal_to: 18,
    message: 'A person must be over 18',
    full_message_format: '%{message}'
  }
end

person = Person.new(age: 17).tap(&:valid?)

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"]

# works with Errors#add
person = Person.new
person.errors.add(:name, full_message: 'A person name cannot be blank')
person.errors.add(:age, 'A person must be over 18', full_message_format: '%{message}')

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"]
```

### Why?

I work with clients that want validation steps or custom error messages that don't necessarily start with the attribute name. Sometimes I wish I had the flexibility of `:base` errors for an attribute without having to write down custom validation methods and keep a consistent way to display error messages on the view.

### Other

Haven't benchmarked it.

As I'm writing the description, I realise people have tried to do this for a while now:

* https://github.com/rails/rails/pull/12457#issuecomment-70334857
* https://github.com/rails/rails/pull/32919

And we can override the format with I18n:
* https://github.com/rails/rails/pull/32956


### Test

You can copy-paste this gist to test it locally.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "alexb52/rails", branch: "full_message_updates"
  gem "byebug"
end

require "active_model"
require "minitest/autorun"
require "logger"
require "byebug"

class Person
  include ActiveModel::Validations

  attr_accessor :name, :favourite_colour, :legacy_code

  validates :name, presence: true # default
  validates :favourite_colour, inclusion: { in: ["blue", "red"], message: "A person favourite colour must be blue or red", full_message_format: '%{message}' }
  validates :legacy_code, format: { with: /\A[a-zA-Z]+\z/, full_message: "PDLP must only include letters" }
end

class PersonTest < Minitest::Test
  def setup
    @errors = Person.new.tap(&:valid?).errors
  end

  def test_full_messages
    assert_equal @errors.full_messages, [
      "Name can't be blank",
      "A person favourite colour must be blue or red",
      "PDLP must only include letters"
    ]
  end

  def test_full_messages_for
    assert_equal @errors.full_messages_for(:name),             ["Name can't be blank"]
    assert_equal @errors.full_messages_for(:favourite_colour), ["A person favourite colour must be blue or red"]
    assert_equal @errors.full_messages_for(:legacy_code),      ["PDLP must only include letters"]
  end

  def test_message
    assert_equal @errors.where(:name).map(&:message),             ["can't be blank"]
    assert_equal @errors.where(:favourite_colour).map(&:message), ["A person favourite colour must be blue or red"]
    assert_equal @errors.where(:legacy_code).map(&:message),      ["PDLP must only include letters"]
  end

  def test_errors_add
    person = Person.new
    person.errors.add(:favourite_colour, full_message: "A person favourite colour must be blue or red")
    assert_equal person.errors.full_messages, ["A person favourite colour must be blue or red"]
  end
end
```

Commit:
Create message_format & full_message validation option

Add the ability to change the message_format
on ActiveModel::Validations full_messages.
Like error.add(:base, 'message') but for any attribute.

Examples:

    class Person
      include ActiveModel::Model

      validates :name,
        presence: { full_message: 'Person requires a name' }
    end

OR

    class Person
      include ActiveModel::Model

      validates :name,
        presence: {
          message: 'Person requires a name',
          message_format: "%{message}"
        }
    end

This would result in

    person = Person.new(name: nil)
    person.valid?

    puts person.errors.full_messages
    => ['Person requires a name']

Amns45 reviewed (changes requested):
class Person
  include ActiveModel::Model
  attr_accessor :age, :name

  validates :name, presence: { full_message: 'A person name cannot be blank' }
  validates :age, numericality: {
                                  greater_than_or_equal_to: 18,
                                  message: 'A person must be over 18',
                                  full_message_format: '%{message}'
                                }
end

person = Person.new(age: 17).tap(&:valid?)

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"]

# works with Errors#add

person = Person.new
person.errors.add(:name, full_message: 'breannawalker)
person.errors.add(:age, 'A person full_message_format: '%{message}')

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"]

Amns45 commented:
class Person
  include ActiveModel::Model
  attr_accessor :age, :name

  validates :name, presence: { full_message: 'A person name cannot be blank' }
  validates :age, numericality: {
                                  greater_than_or_equal_to: 18,
                                  message: 'A person must be over 18',
                                  full_message_format: '%{message}'
                                }
end

person = Person.new(age: 17).tap(&:valid?)

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"]

# works with Errors#add

person = Person.new
person.errors.add(:name, full_message: 'A person name cannot be blank')
person.errors.add(:age, 'A person must be over 18', full_message_format: '%{message}')

person.errors.full_messages
["A person must be over 18", "A person name cannot be blank"]

person.errors.where(:name).map(&:message)
["A person name cannot be blank"][]()**@josh **

rafaelfranca commented:
Code looks good so far but we need to document those two new options. Can you update the PR to add documentation?

AlexB52 commented:
@rafaelfranca Sorry I somehow missed the comments above, I will add documentation

Commit:
Add API doc for :full_message_format and :full_message options

Commit:
Update ActiveRecord validations documentation

The documentation now provides descriptions for both :full_message
and full_message_format options.

AlexB52 commented:
@rafaelfranca is that what you had in mind for documentation updates?

Commit:
Switch to full_message_format option

Commit:
Add format full validation messages section on ActiveRecord validations guides

AlexB52 commented on guides/source/active_record_validations.md:
````diff
@@ -1372,6 +1454,99 @@ irb> person.errors.size
+```
+
+```yml
+activerecord:
+  errors:
+    models:
+      person:
+        attributes:
+          format: "** %{attribute} %{message} **"
+          name:
+            format: "-- %{message} --"
+            blank: "can't be blank"
+          age:
+            not_a_number: "is not a number"
````
This requires `config.active_model.i18n_customize_full_message = true` in application.rb file.
